messaging/broadcast.go

```diff
@@ -64,28 +69,35 @@ func (b *Broadcast) handleContactNodes(msg struct {
 	*onet.TreeNode
 	ContactNodes
 }) error {
-	log.Lvl3(b.Info(), "Received message from", msg.TreeNode.String())
+	log.Lvl3(b.ServerIdentity(), "Received contact message from", msg.TreeNode.ServerIdentity)
+	err := b.SendTo(msg.TreeNode, &Done{})
+	if err != nil {
+		return err
+	}
 	if msg.TreeNode.ID.Equal(b.Root().ID) {
-		b.repliesLeft = len(b.Tree().List()) - b.tnIndex - 1
-		if b.repliesLeft == 0 {
-			log.Lvl3("Won't contact anybody - finishing")
-			b.SendTo(b.Root(), &Done{})
-			return nil
-		}
-		log.Lvl3(b.Info(), "Contacting nodes:", b.repliesLeft)
-		// Connect to all nodes that are later in the TreeNodeList, but only if
-		// the message comes from root
+		// Connect to all nodes that are later in the TreeNodeList.
 		for _, tn := range b.Tree().List()[b.tnIndex+1:] {
-			log.Lvl3("Connecting to", tn.String())
+			log.Lvl3("Contacting", tn.String())
 			err := b.SendTo(tn, &ContactNodes{})
 			if err != nil {
-				return nil
+				return err
 			}
 		}
-	} else {
-		// Tell the caller we're done
-		log.Lvl3("Sending back to", msg.TreeNode.ServerIdentity.String())
-		b.SendTo(msg.TreeNode, &Done{})
+	}
+	b.contactRcv--
+	return b.verifyDone()
+}
+
+func (b *Broadcast) verifyDone() error {
+	if b.contactRcv+b.doneRcv == 0 {
+		err := b.SendTo(b.Root(), &Done{})
+		if err != nil {
+			log.ErrFatal(err)
```

Comment:
Wait, what? A failed sendTo somewhere down deep inside of some protocol's broadcasting attempt can fill the server? That can't be right...

Reply:
No, it can't be right. Thanks for the catch!